### PR TITLE
Issue 32 jsoniq fn

### PR DESCRIPTION
**Pull request by CanBerker**
All jsoniq specific functions except null(), encode...() and decode...() have been implemented.

### CODE REVIEW

File: src/main/resources/test_files/runtime/FunctionObjectRemoveKeys.iq [hunk 1]
```diff
@@ -0,0 +1,2 @@
+(:JIQS: ShouldRun; Output="( { "Engineer" : "Scott" } )" :)
+let $o := {"Captain" : "Kirk", "First Officer" : "Spock", "Engineer" : "Scott"} return remove-keys($o, ("Captain", "First Officer"))
```
Comment (ghislainfourny) [inline src/main/resources/test_files/runtime/FunctionObjectRemoveKeys.iq]: We'd need a test that remove non-existing keys, and another that removes a mix of existing and non-existing keys.

File: src/main/resources/test_files/runtime/FunctionObjectProject2.iq [hunk 2]
```diff
@@ -0,0 +1,2 @@
+(:JIQS: ShouldRun; Output="( { } )" :)
```
Comment (ghislainfourny) [inline src/main/resources/test_files/runtime/FunctionObjectProject2.iq]: We'd need a test that projects a mix of existing and non-existing keys.

File: src/main/resources/test_files/runtime/FunctionObjectIntersect.iq [hunk 3]
```diff
@@ -0,0 +1,2 @@
+(:JIQS: ShouldRun; Output="( { "a" : [ abc, 2 ], "b" : [ 2, ab ], "c" : [ [ 1, 2 ], 5e+1 ] } )" :)
```
Comment (ghislainfourny) [inline src/main/resources/test_files/runtime/FunctionObjectIntersect.iq]: Just a side note: in the long term, the syntax for expected results may need to be improved. Here you can see that this is a quoted string containing quotes. No need to address this now though if it works as is, no worries.

File: src/main/resources/test_files/runtime/FunctionObjectDescendantPairs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+(:JIQS: ShouldRun; Output="( { "a" : [ 1, { "b" : 2 } ] }, { "b" : 2 }, { "o" : { "c" : 3 } }, { "c" : 3 } )" :)
```
Comment (ghislainfourny) [inline src/main/resources/test_files/runtime/FunctionObjectDescendantPairs]: "d" instead of "o"?

File: src/main/resources/test_files/runtime/FunctionObjectDescendantPairs [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+let $o := ({ "a" : [1, {"b" : 2}], "d" : {"c" : 3} }) return descendant-pairs($o)
+
+
+
+(: Although the code works properly, this test file doesn't execute bc of an issue I haven't solved.
+testRuntimeIterators[53:/home/travis/build/Sparksoniq/sparksoniq/src/main/resources/test_files/runtime/FunctionObjectDescendantPairs.iq](iq.RuntimeTests)  Time elapsed: 0.142 sec  <<< ERROR!
+org.apache.spark.SparkException: Job aborted due to stage failure: Exception while getting task result: com.esotericsoftware.kryo.KryoException: Unable to find class: '*(
```
Comment (ghislainfourny) [inline src/main/resources/test_files/runtime/FunctionObjectDescendantPairs]: This may be a known issue (see other PRs and bugs). It would be nice to reproduce it with code as small as possible, but in a different PR.

File: src/main/resources/test_files/runtime/FunctionObjectDescendant.iq [hunk 6]
```diff
@@ -0,0 +1,4 @@
+(:JIQS: ShouldRun; Output="( { "a" : [ 1, { "b" : 2 } ], "o" : { "c" : 3 } }, { "b" : 2 }, { "c" : 3 } )" :)
```
Comment (ghislainfourny) [inline src/main/resources/test_files/runtime/FunctionObjectDescendant.iq]: We need tests on corner cases (top-level array, empty object, empty array, list of atomics, etc)

File: src/main/resources/test_files/runtime/FunctionObjectAccumulate.iq [hunk 7]
```diff
@@ -0,0 +1,3 @@
+(:JIQS: ShouldRun; Output="( { "b" : [ 2, [ 2, xyz ] ], "c" : [ 3, { "d" : 200.0 } ] } )" :)
```
Comment (ghislainfourny) [inline src/main/resources/test_files/runtime/FunctionObjectAccumulate.iq]: We need tests with keys that do not appear in all objects, as well as with one object, zero objects and three objects.

File: src/main/resources/test_files/runtime/FunctionArrayFlatten.iq [hunk 8]
```diff
@@ -0,0 +1,2 @@
+(:JIQS: ShouldRun; Output="( 0, x, { "a" : [ 1, { "b" : 2 } ] }, 3, c, d )" :)
+let $o := ([0, "x", { "a" : [1, {"b" : 2}] }, [3, ["c", "d"]]]) return flatten($o)
```
Comment (ghislainfourny) [inline src/main/resources/test_files/runtime/FunctionArrayFlatten.iq]: We need tests on corner cases (empty array, flat array, ...)

File: src/main/resources/test_files/runtime/FunctionArrayDescendant.iq [hunk 9]
```diff
@@ -0,0 +1,4 @@
+(:JIQS: ShouldRun; Output="( [ 0, x, { "a" : [ 1, { "b" : 2 }, [ 2.5 ] ], "o" : { "c" : 3 } } ], [ 1, { "b" : 2 }, [ 2.5 ] ], [ 2.5 ])" :)
+let $o := ([0, "x", { "a" : [1, {"b" : 2}, [2.5]], "o" : {"c" : 3} }]) return descendant-arrays($o)
```
Comment (ghislainfourny) [inline src/main/resources/test_files/runtime/FunctionArrayDescendant.iq]: We need to test corner cases (top-level object, flat array, empty array, empty object)
